compiler/rustc_middle/src/ty/list.rs

```diff
@@ -210,23 +213,216 @@ unsafe impl<T: Sync> Sync for List<T> {}
 
 // We need this since `List` uses extern type `OpaqueListContents`.
 #[cfg(parallel_compiler)]
-use rustc_data_structures::sync::DynSync;
-
-use super::TyCtxt;
-#[cfg(parallel_compiler)]
 unsafe impl<T: DynSync> DynSync for List<T> {}
 
 // Safety:
-// Layouts of `Equivalent<T>` and `List<T>` are the same, modulo opaque tail,
-// thus aligns of `Equivalent<T>` and `List<T>` must be the same.
+// Layouts of `ListSkeleton<T>` and `List<T>` are the same, modulo opaque tail,
+// thus aligns of `ListSkeleton<T>` and `List<T>` must be the same.
 unsafe impl<T> Aligned for List<T> {
-    const ALIGN: ptr::Alignment = {
+    const ALIGN: ptr::Alignment = align_of::<ListSkeleton<T>>();
+}
+
+/// A [`List`] that additionally stores type information inline to speed up
+/// [`TypeVisitableExt`](super::TypeVisitableExt) operations.
+#[repr(C)]
+pub struct ListWithCachedTypeInfo<T> {
+    skel: ListWithCachedTypeInfoSkeleton<T>,
+    opaque: OpaqueListContents,
```

Comment:
why is that moved out of `ListSkeleton`?

Reply:
Because accessing a field with an extern type tail will always panic, because we don't know the align of the extern type and therefore the padding before the field. So basically if we just put `List` as the tail of `ListWithCachedTypeInfo`, then any attempt to access the list will panic.

Reply:
Hmmm apparently there is a special case if the field with an extern type tail is the only field. Which i guess makes sense since there can't be any padding in that case. If you change the outer type to have some extra data, your example will panic: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=178bcc676aa494d0452be7b3baeb5d83

> I don't think we should ever copy `ListSkeleton` to the stack

I agree. If my code does that it's not intended.

Reply:
I've added some comments to the `*Skeleton` types to hopefully clear this up a bit.

Relevant diff: https://github.com/rust-lang/rust/compare/c4f1f7aa586451154c669711abc9c4d9ad9b1075..fcc477fbd0871d8fab045bd7e9524172ab10e51c